core/services/relay/evm/binding.go

```diff
@@ -0,0 +1,145 @@
+package evm
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+)
+
+type binding interface {
+	GetLatestValue(ctx context.Context, params any) ([]byte, error)
+	Bind(binding commontypes.BoundContract) error
+	SetCodec(codec commontypes.Codec)
+	Register() error
+	Unregister() error
+}
+
+type methodBinding struct {
+	address      common.Address
+	contractName string
+	method       string
+	client       evmclient.Client
+	codec        commontypes.Codec
+}
+
+var _ binding = &methodBinding{}
+
+func (m *methodBinding) SetCodec(codec commontypes.Codec) {
+	m.codec = codec
+}
+
+func (m *methodBinding) Register() error {
+	return nil
+}
+
+func (m *methodBinding) Unregister() error {
+	return nil
+}
+
+func (m *methodBinding) GetLatestValue(ctx context.Context, params any) ([]byte, error) {
+	data, err := m.codec.Encode(ctx, params, wrapItemType(m.contractName, m.method, true))
+	if err != nil {
+		return nil, err
+	}
+
+	callMsg := ethereum.CallMsg{
+		To:   &m.address,
+		From: m.address,
+		Data: data,
+	}
+
+	return m.client.CallContract(ctx, callMsg, nil)
+}
+
+func (m *methodBinding) Bind(binding commontypes.BoundContract) error {
+	m.address = common.HexToAddress(binding.Address)
+	return nil
+}
+
+type eventBinding struct {
+	address      common.Address
+	contractName string
+	eventName    string
+	lp           logpoller.LogPoller
+	hash         common.Hash
+	codec        commontypes.Codec
+	pending      bool
+	subscribed   bool
+	hasAddr      bool
+}
+
+func (e *eventBinding) SetCodec(codec commontypes.Codec) {
+	e.codec = codec
+}
+
+func (e *eventBinding) Register() error {
+	e.subscribed = true
+	if !e.hasAddr {
+		return nil
```

Comment:
we can't retrieve an event without contract address, should this retrun an  error instead?

Reply:
I wanted you to be able to start the `ChainReader` without needing to worry if the bindings have been added already or will be added later by a user.

Reply:
Doing it like this we risk accidentally trying to retrieve logs from 0x0 address, which will alway return not found and may be confusing to debug in development. Maybe we should add a debug log in GetLatestValue for events that notices that 0x0000 is being used

Reply:
I can return an error if it wasn't initialized in the call.  Otherwise, we need to enforce that whenever it's started bindings are always known and I'm not sure that'll be the case.